web/packages/teleport/src/Roles/RoleEditor/standardmodel.ts

```diff
@@ -211,25 +409,52 @@ export function roleEditorModelToRole(roleModel: RoleEditorModel): Role {
   };
 
   for (const spec of roleModel.accessSpecs) {
-    if (spec.kind === 'node') {
-      const labels = {};
-      for (const { name, value } of spec.labels) {
-        if (!Object.hasOwn(labels, name)) {
-          labels[name] = value;
-        } else if (typeof labels[name] === 'string') {
-          labels[name] = [labels[name], value];
-        } else {
-          labels[name].push(value);
-        }
+    const { kind } = spec;
+    switch (kind) {
+      case 'node': {
+        role.spec.allow.node_labels = labelsModelToLabels(spec.labels);
+        role.spec.allow.logins = spec.logins.map(opt => opt.value);
+        break;
+      }
+      case 'kube_cluster': {
+        role.spec.allow.kubernetes_groups = spec.groups.map(opt => opt.value);
+        role.spec.allow.kubernetes_labels = labelsModelToLabels(spec.labels);
+        role.spec.allow.kubernetes_resources = spec.resources.map(
+          ({ kind, name, namespace, verbs }) => ({
+            kind: kind.value,
+            name,
+            namespace,
+            verbs: verbs.map(opt => opt.value),
+          })
+        );
+        break;
       }
-      role.spec.allow.node_labels = labels;
-      role.spec.allow.logins = spec.logins.map(opt => opt.value);
+      default:
+        kind satisfies never;
```

Comment:
maybe I'm overlooking something, but does this have any effect?

Reply:
@kiosion Yes, that's a very useful pattern that I learned a couple of months ago: `kind satisfies never` means that at this point, the `kind` variable cannot possibly hold any value. Using it in a `default` block of a `case` statement means that (at least from the TS Compiler's perspective) every possible case has been handled. Thus, adding another possibility to the union type of `kind` would cause a compiler error and enforce handling this case before the code can be merged.

Reply:
(That's one of the things that I've been missing from languages like Rust, and I'm very happy that there's a way to express it in TS.)